apm-sniffer/bytebuddy-patch/src/test/java/org/apache/skywalking/apm/agent/bytebuddy/SWClassFileLocator.java

```diff
@@ -0,0 +1,174 @@
+                } catch (Throwable e) {
+                    LOGGER.error(e, "Resolve bytecode of class failure");
+                }
+            }
+        }, "SWClassFileLocator");
+        thread.setDaemon(true);
+        thread.start();
+    }
+
+    @Override
+    public Resolution locate(String name) throws IOException {
+        if (!match(name)) {
+            return new Resolution.Illegal(name);
+        }
+        if (closed) {
+            throw new IOException("resolve class failure: closed");
+        }
+        // get class binary representation in a clean thread, avoiding nest calling transformer!
+        ResolutionFutureTask futureTask = new ResolutionFutureTask(name);
+        queue.offer(futureTask);
+        try {
+            return futureTask.getFuture().get(timeoutSeconds, TimeUnit.SECONDS);
+        } catch (Exception e) {
+            throw new IOException("resolve class failure: " + name, e);
+        }
+    }
+
+    private boolean match(String name) {
+        boolean matched = false;
+        for (String typeNameTrait : typeNameTraits) {
+            if (name.contains(typeNameTrait)) {
+                matched = true;
+                break;
+            }
+        }
+        return matched;
+    }
+
+    private Resolution getResolution(String name) throws Exception {
+        SWExtractionClassFileTransformer classFileTransformer = new SWExtractionClassFileTransformer(name);
+        try {
+            instrumentation.addTransformer(classFileTransformer, true);
+            Class aClass = locateClass(name);
+            if (aClass == null) {
+                return new Resolution.Illegal(name);
+            }
+            instrumentation.retransformClasses(new Class[]{aClass});
+        } finally {
+            instrumentation.removeTransformer(classFileTransformer);
+        }
+
+        return classFileTransformer.getBinaryRepresentation() != null ?
+                new Resolution.Explicit(classFileTransformer.getBinaryRepresentation()) :
+                new Resolution.Illegal(name);
+    }
+
+    private Class locateClass(String className) {
+        // find class in classloader
+        try {
+            return classLoadingDelegate.locate(className);
+        } catch (ClassNotFoundException e) {
+        }
+
+        // find class in instrumentation
+        Class[] allLoadedClasses = instrumentation.getAllLoadedClasses();
+        for (int i = 0; i < allLoadedClasses.length; i++) {
+            Class aClass = allLoadedClasses[i];
+            if (className.equals(aClass.getName())) {
+                return aClass;
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public void close() throws IOException {
+        closed = true;
+        queue.clear();
+        thread.interrupt();
+    }
+
+    public int getTimeoutSeconds() {
+        return timeoutSeconds;
+    }
+
+    public void setTimeoutSeconds(int timeoutSeconds) {
+        this.timeoutSeconds = timeoutSeconds;
+    }
```

Comment:
Unused methods.

Reply:
It's used to config resolve timeout, not every public method need a usage, might need it somewhere later.

Reply:
This is only a test code, I think usually manual hard code is good enough?